pyro/distributions/relaxed_categorical.py

```diff
@@ -0,0 +1,155 @@
+
+    def _process_data(self, x):
+        if x is not None:
+            if isinstance(x, list):
+                x = np.array(x)
+            elif not isinstance(x, (Variable, torch.Tensor, np.ndarray)):
+                raise TypeError(("Data should be of type: list, Variable, Tensor, or numpy array"
+                                 "but was of {}".format(str(type(x)))))
+        return x
+
+    def batch_shape(self, x=None):
+        """
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.batch_shape`
+        """
+        event_dim = 1
+        ps = self.ps
+        if x is not None:
+            x = self._process_data(x)
+            x_shape = x.shape if isinstance(x, np.ndarray) else x.size()
+            try:
+                ps = self.ps.expand(x_shape[:-event_dim] + self.event_shape())
+            except RuntimeError as e:
+                raise ValueError("Parameter `ps` with shape {} is not broadcastable to "
+                                 "the data shape {}. \nError: {}".format(ps.size(), x.size(), str(e)))
+        return ps.size()[:-event_dim]
+
+    def event_shape(self):
+        """
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.event_shape`
+        """
+        event_dim = 1
+        return self.ps.size()[-event_dim:]
+
+    def shape(self, x=None):
+        """
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.shape`
+        """
+        if self.one_hot:
+            return self.batch_shape(x) + self.event_shape()
+        return self.batch_shape(x) + (1,)
+
+    def sample(self):
+        """
+        Draws either a single sample (if self.logits.dim() == 1), or one sample per param (if self.logits.dim() == 2).
+        Reparameterized.
+        """
+
+        # Sample Gumbels, G_k = -log(-log(U))
+        uniforms = torch.zeros(self.logits.data.size()).uniform_()
+        eps = _get_clamping_buffer(uniforms)
+        uniforms = uniforms.clamp(min=eps, max=1 - eps)
+        gumbels = Variable(uniforms.log().mul(-1).log().mul(-1))
+
+        # Reparameterize
+        z = F.log_softmax((self.logits + gumbels) / self.temperature)
+        return z if self.reparameterized else z.detach()
+
+    def batch_log_pdf(self, x):
+        """
+        Evaluates log probability densities for one or a batch of samples and parameters.
+
+        :return: tensor with log probabilities for each of the batches.
+        :rtype: torch.autograd.Variable
+        """
+        n = self.event_shape()[0]
+        logits = self.logits.expand(self.shape(x))
+        log_scale = Variable(log_gamma(torch.Tensor([n]).expand(
+            self.shape(x)))) - self.temperature.log().mul(-(n - 1))
+        scores = logits.log() + x.mul(-self.temperature)
+        scores = scores.sum(dim=-1)
+
+        log_part = n * logits.mul(x.mul(-self.temperature).exp()).sum(dim=-1).log()
+        return (scores - log_part + log_scale).contiguous()
+
+
+class Exp(Bijector):
```

Comment:
I'd rename this to `ExpBijector` and move it into `transformed_distributions.py`, since there are many other uses.